File: Labs/Lab 8/Main.java
public int coinChange(int[] coins, int amount) {
    int[] dp = new int[amount + 1];
    Arrays.fill(dp, amount + 1); // Initialize with a large value, indicating 'impossible' 
    dp[0] = 0; // Base case: 0 coins needed for amount 0

    for (int i = 1; i <= amount; i++) {
        for (int coin : coins) {
            if (coin <= i) {
                dp[i] = Math.min(dp[i], 1 + dp[i - coin]); 
            }
        }
    }

   return dp[amount] > amount ? -1 : dp[amount];
}

File: Labs/Lab 8/Main.py
def coinChange(self, coins: List[int], amount: int) -> int:
    dp = [amount + 1] * (amount + 1)  # Initialize with a large value
    dp[0] = 0  # Base case

    for i in range(1, amount + 1):
        for coin in coins:
            if coin <= i:
                dp[i] = min(dp[i], 1 + dp[i - coin])

    return dp[amount] if dp[amount] <= amount else -1
